File: packages/python/cronixui/list.py
"""CronixUI List Component"""

from typing import Optional, Callable
from .core import create_el


class List:
    """List component."""

    def __init__(self, items: list, clickable: bool = False):
        self.items = items
        self.clickable = clickable
        self.element = self._render()

    def _render(self):
        el = create_el("div", "cn-list")
        for item in self.items:
            item_el = create_el("div", "cn-list-item")
            if self.clickable:
                item_el.classList.add("cn-list-item-clickable")

            if isinstance(item, dict):
                if icon := item.get("icon"):
                    icon_el = create_el("span", "cn-list-item-icon")
                    icon_el.innerHTML = icon
                    item_el.appendChild(icon_el)

                content = create_el("div", "cn-list-item-content")
                if title := item.get("title"):
                    title_el = create_el("div", "cn-list-item-title")
                    title_el.textContent = title
                    content.appendChild(title_el)

                if subtitle := item.get("subtitle"):
                    subtitle_el = create_el("div", "cn-list-item-subtitle")
                    subtitle_el.textContent = subtitle
                    content.appendChild(subtitle_el)

                item_el.appendChild(content)

                if actions := item.get("actions"):
                    actions_el = create_el("div", "cn-list-item-actions")
                    actions_el.innerHTML = actions
                    item_el.appendChild(actions_el)
            else:
                content = create_el("div", "cn-list-item-content")
                content.textContent = str(item)
                item_el.appendChild(content)

            el.appendChild(item_el)
        return el

File: packages/rust/cronixui/src/components/list.rs
//! List component

pub struct ListItem {
    pub title: String,
    pub subtitle: Option<String>,
}

impl ListItem {
    pub fn new(title: impl Into<String>) -> Self {
        Self { title: title.into(), subtitle: None }
    }

    pub fn subtitle(mut self, subtitle: impl Into<String>) -> Self {
        self.subtitle = Some(subtitle.into());
        self
    }
}

pub struct List {
    pub items: Vec<ListItem>,
}

impl List {
    pub fn new() -> Self {
        Self { items: Vec::new() }
    }

    pub fn item(mut self, item: ListItem) -> Self {
        self.items.push(item);
        self
    }
}

impl Default for List {
    fn default() -> Self {
        Self::new()
    }
}
